webapp/src/app/shared/phaser/coords-reporter.ts

```diff
@@ -0,0 +1,65 @@
+import { Point } from '../../models/cross-code-map';
+import { EditorView } from '../../models/editor-view';
+import { Globals } from '../globals';
+
+export class CoordsReporter extends Phaser.GameObjects.GameObject {
+	private rawCoords: Point = { x: 0, y: 0 };
+
+	private get coords(): Point {
+		return {
+			x: this.rawCoords.x,
+			y: this.rawCoords.y + this.entityLevel,
+		};
+	}
+
+	private get isEntityMode() {
+		return (
+			Globals.globalEventsService.currentView.value ===
+			EditorView.Entities
+		);
+	}
+
+	private get entityLevel() {
+		const maybeLevel =
+			Globals.globalEventsService.selectedEntity.value?.details.level;
+
+		return maybeLevel
+			? Globals.map.levels[maybeLevel.level].height + maybeLevel.offset
+			: 0;
+	}
+
+	// used to help hide coords without firing duplicate events
+	private lastModeWasTile = true;
+
+	private preUpdate() {
+		const tileMode = !this.isEntityMode;
+		if (tileMode) {
+			if (!this.lastModeWasTile) {
+				Globals.globalEventsService.updateCoords.next(null);
+			}
+
+			return;
+		}
+
+		const pointer = this.scene.input.activePointer;
+
+		const newRaw = {
+			x: Math.floor(pointer.worldX),
+			y: Math.floor(pointer.worldY),
+		};
+
+		// don't send events if we don't have to
+		if (newRaw.x === this.rawCoords.x && newRaw.y === this.rawCoords.y) {
+			return;
+		}
+
+		this.rawCoords = newRaw;
+		this.lastModeWasTile = tileMode;
+
+		// send event
+		Globals.globalEventsService.updateCoords.next({
+			...this.coords,
+			z: this.entityLevel,
+		});
+	}
```

Comment:
subscribe to subjects instead of polling the view every frame. The preUpdate function should always report the coordinates. If the view changes the actual game object can be disabled.
Something like this:

```typescript
import { Point } from "../../models/cross-code-map";
import { EditorView } from "../../models/editor-view";
import { Globals } from "../globals";
import { Subscription } from "rxjs";
import { BaseObject } from "./base-object";

export class CoordsReporter extends BaseObject {
  private rawCoords: Point = { x: 0, y: 0 };

  private sub!: Subscription;

  constructor(scene: Phaser.Scene, active = true) {
    super(scene, "coordsReporter", active);
  }

  protected init(): void {
    this.sub = Globals.globalEventsService.currentView.subscribe((view) => {
      this.setActive(view === EditorView.Entities);
    });
  }

  protected activate(): void {}

  protected deactivate(): void {
    Globals.globalEventsService.updateCoords.next();
  }

  destroy(fromScene?: boolean) {
    this.sub.unsubscribe();
    super.destroy(fromScene);
  }

  private get coords(): Point {
    return {
      x: this.rawCoords.x,
      y: this.rawCoords.y + this.entityLevel,
    };
  }

  private get entityLevel() {
    const maybeLevel =
      Globals.globalEventsService.selectedEntity.value?.details.level;

    return maybeLevel
      ? Globals.map.levels[maybeLevel.level].height + maybeLevel.offset
      : 0;
  }

  preUpdate(time: number, delta: number) {
    const pointer = this.scene.input.activePointer;

    const newRaw = {
      x: Math.floor(pointer.worldX),
      y: Math.floor(pointer.worldY),
    };

    // don't send events if we don't have to
    if (newRaw.x === this.rawCoords.x && newRaw.y === this.rawCoords.y) {
      return;
    }

    this.rawCoords = newRaw;

    // send event
    Globals.globalEventsService.updateCoords.next({
      ...this.coords,
      z: this.entityLevel,
    });
  }
}

```
I also added the `BaseObject` class to have proper typings for preUpdate